client/main.go

```diff
@@ -0,0 +1,1854 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gookit/color"
+)
+
+const (
+	Version          = "3.12.0.240221" // Overwritten automatically by -ldflags
+	ReportTimeout    = 3 * time.Minute
+	OAUTH_CLIENT_ID  = "IdFRwa3SGA8eMpzhRVFMg5Ts8sPK93xBjif93x0F"
+	WindowsPathLimit = 250
+
+	// PATHS
+	server_default     = "https://www.blenderkit.com" // default address to production blenderkit server
+	gravatar_dirname   = "bkit_g"                     // directory in safeTempDir() for gravatar images
+	cleanfile_path     = "blendfiles/cleaned.blend"   // relative path to clean blend file in add-on directory
+	upload_script_path = "upload_bg.py"               // relative path to upload script in add-on directory
+
+	// EMOJIS
+	EmoOK            = "✅"
+	EmoCancel        = "⛔"
+	EmoWarning       = "⚠️ " // Needs space at the end for proper alignment, not sure why.
+	EmoInfo          = "ℹ️ "
+	EmoError         = "❌"
+	EmoNetwork       = "📡"
+	EmoNewConnection = "🤝"
+	EmoDisconnecting = "👐"
+	EmoSecure        = "🔒"
+	EmoInsecure      = "🧨"
+)
+
+var (
+	SystemID        *string // Unique ID of the current system (15 integers)
+	PlatformVersion string
+	Port            *string
+	Server          *string
+
+	CodeVerifier    string //Used for OAuth2
+	CodeVerifierMux sync.Mutex
+
+	lastReportAccess    time.Time
+	lastReportAccessMux sync.Mutex
+
+	ActiveAppsMux sync.Mutex
+	ActiveApps    []int
+
+	Tasks                map[int]map[string]*Task
+	TasksMux             sync.Mutex
+	AddTaskCh            chan *Task
+	TaskProgressUpdateCh chan *TaskProgressUpdate
+	TaskMessageCh        chan *TaskMessageUpdate
+	TaskFinishCh         chan *TaskFinish
+	TaskErrorCh          chan *TaskError
+	TaskCancelCh         chan *TaskCancel
+
+	ClientAPI, ClientDownloads, ClientUploads, ClientSmallThumbs, ClientBigThumbs *http.Client
+
+	BKLog   *log.Logger
+	ChanLog *log.Logger
+)
+
+func init() {
+	Tasks = make(map[int]map[string]*Task)
+	AddTaskCh = make(chan *Task, 100)
+	TaskProgressUpdateCh = make(chan *TaskProgressUpdate, 1000)
+	TaskMessageCh = make(chan *TaskMessageUpdate, 1000)
+	TaskFinishCh = make(chan *TaskFinish, 100)
+	TaskCancelCh = make(chan *TaskCancel, 100)
+	TaskErrorCh = make(chan *TaskError, 100)
+
+	PlatformVersion = runtime.GOOS + " " + runtime.GOARCH + " go" + runtime.Version()
```

Comment:
Just letting you know that we are going to use this header on the new Sessions page to display the user's device's OS. I can see it is going to change. I guess, the change is for the better as it can be parsed more easily (just split on space). I guess that reduction in the amount of provided information is not going to be a problem. I still wonder whether it wouldn't be better to split it into three different headers so that the server wouldn't need to parse it at all but I guess that this format is error-proof. But if more information will be included later, then a split to more headers would be welcome. Just letting you know so that you can take the new usage of the header into account i.e. if you had some further planes with it because this seems to be the best time to change things.

Reply:
Oh, I've forgot about this part of the code. This was just quick implementation. If we are going to use this string as identifier of the device, the go version does not make much sense (we will know which version of Go we have used to compile for the addon-version).

I like the idea of splitting the informaton into more headers and making the final implementation now. Which fields we would like?

- platform-architecture - I think it is good to have debug information about whether amd64(x86_64) or arm64 machine was used. Blender does not support 32bit systems AFAIU.
- platform-os - also handy for debugging, I would suggest to rename the darwin to macos, for clarity to users.
- platform-os-version - I am not sure how to get the `14.2.1` from python's platform.platform() `macOS-14.2.1-arm64-arm-64bit` robustly in Go. I can get this from the add-on, or I can try some other library, but maybe it is not that crucial. Also getting the info from the add-on might not be working if we later try to add Godot support, which programs add-on in C#.

I would add also:
- client-version: it would be mostly the same as addon-version, but in some cases it could differ - user starts Blender with BlenderKit v.3.22.0 lets say, and then older Blender with v3.20.0 - it will reuse the running client.

If we plan to support Godot and other 3D softwares in future, maybe a `addon-type` would be useful. For now we have only `blender`, but later we could have also: `godot`, `unity`, or even `ableton` if we ever start expanding to sound fields.

And what about this? https://github.com/BlenderKit/blenderkit/issues/1017 Seems like good time to also solve this problem. So maybe something like: `system-nickname` would be ok? We can pre-generate that in add-on preferences, allow users to change it and allow them to disable this header.

"Platform-Arch": "arm64" / "amd64"
"Platform-OS": "windows" / "macOS" / "linux"
"system-id": 116830648666783,
"system-nickname": "Meaningfull Panda"
"addon-type": "blenderkit-blender"
"addon-version": X.Y.Z.YYMMDDD
"client-version": X.Y.Z.YYMMDD

Should I keep the `platform-version` for backward compatibility? WDYT @PetrDlouhy @radekholy24 @vilemduha

Reply:
Well, does someone understand why they do not translate "Darwin" to "macOS" in Python? https://bugs.python.org/issue35516 :sweat_smile:

Reply:
Actually, I don't understand the difference between Darwin and macOS but the translation does not seem to be one-to-one? https://github.com/python/cpython/blob/main/Lib/platform.py#L1257-L1262

Reply:
As for the `addon-type`, I'd vote for adding it once we add support for another 3D software. I don't think we need to define what is obvious. But I have no strong opinion on that.

Reply:
@radekholy24 Personally I do not know anybody running e.g. https://www.puredarwin.org/ (https://github.com/apple/darwin-xnu, https://github.com/apple/xnu). Even though Go binaries targeted to GOOS=darwin can be probably run on Darwin also, most Go projects just refer to it as macOS.

For now I can just take the value from the Python in add-on starting the Client. In the future however this will need change as maybe Godot C# will not be able to detect Darwin/MacOS difference that easily. In that case implementing https://github.com/python/cpython/blob/main/Lib/platform.py#L464 in Go in Client would be an option.

Reply:
To make it clear, my position on splitting `Platform-Version` into separate headers depends on how reliable would parsing of the header be. As long as it is just a string composed of three substrings (hopefully, not containing a space) concatenated with a space, I slightly prefer a single header. However, once it becomes too complex to encode all information into a single header (possibly, requiring escaping stuff, etc.), then I am for splitting it into multiple headers. So the question here is if OS+arch(+go) is enough for us middle term.

Reply:
I am not aware of any possible usage of the OS _version_. I am not going to include that on the new Sessions page as none of the big names on the internet does.

Reply:
While I am not going to include the system architecture on that page either, I agree that this information might be useful for debugging.

Reply:
Yeah, if different versions of the addon can (re)use different versions of the client, I'm definitely for including that information. (And also adocating for a proper versioning scheme independent of the addon's versioning scheme but that is a different discussion.)

Reply:
BTW, since the only _version_ in the new `Platform-Version` is the version of Go, I think it does not make any sense to call it `*-Version` anymore. And since the values change as well anyway. So maybe call it just `Platform` or whatever fits better. (But for compatibility reasons, copy it also to the `Platform-Version` header for some time.)

Reply:
And my (probably) last thought on this, it just came to my mind that we can encode those possible compound headers the same way as `Set-Cookie` or `Content-Type`. I mean something like `Platform: arch=x86; os=Linux; go=go1.2.3`. But since Python deprecated the [support for parsing such header formats](https://docs.python.org/3/library/cgi.html#cgi.parse_header), it would probably just complicate things... I just wanted to leave it here... :sweat_smile:

Reply:
Oh, and... As for whether to send `macOS` or `darwin` or both, what for other systems... Again, I don't know. Using sending something what [`platform.system_alias`](https://docs.python.org/3/library/platform.html#platform.system_alias) understands would be convenient for server but might be difficult for the client. If there is a portable standard on platform identifiers, let's use it. (I am not aware of any portable one.) If not, let's use whatever is the easiest way.

Reply:
BTW, the current daemon sends the Python version in the `User-Agent` header. Although, I am sure it is not intentional...